plaidml/bridge/openvino/ops/proposal.cpp

```diff
@@ -0,0 +1,369 @@
+      box_y0 = op::clip(box_y0, edsl::Tensor(0), edsl::Tensor(img_H));
+      box_x1 = op::clip(box_x1, edsl::Tensor(0), edsl::Tensor(img_W));
+      box_y1 = op::clip(box_y1, edsl::Tensor(0), edsl::Tensor(img_H));
+    }
+
+    auto box_width = box_x1 - box_x0 + coordinates_offset;
+    auto box_height = box_y1 - box_y0 + coordinates_offset;
+    auto box_ctr_x = box_x0 + box_width * 0.5f;
+    auto box_ctr_y = box_y0 + box_height * 0.5f;
+
+    auto pred_ctr_x = box_ctr_x + dx * box_width;
+    auto pred_ctr_y = box_ctr_y + dy * box_height;
+    auto pred_width = edsl::exp(d_log_w) * box_width;
+    auto pred_height = edsl::exp(d_log_h) * box_height;
+
+    box_x0 = pred_ctr_x - pred_width * 0.5f;
+    box_y0 = pred_ctr_y - pred_height * 0.5f;
+    box_x1 = pred_ctr_x + pred_width * 0.5f;
+    box_y1 = pred_ctr_y + pred_height * 0.5f;
+
+    if (attrs.clip_before_nms) {
+      box_x0 = op::clip(box_x0, edsl::Tensor(0), edsl::Tensor(img_W - coordinates_offset));
+      box_y0 = op::clip(box_y0, edsl::Tensor(0), edsl::Tensor(img_H - coordinates_offset));
+      box_x1 = op::clip(box_x1, edsl::Tensor(0), edsl::Tensor(img_W - coordinates_offset));
+      box_y1 = op::clip(box_y1, edsl::Tensor(0), edsl::Tensor(img_H - coordinates_offset));
+    }
+
+    auto new_box_width = box_x1 - box_x0 + coordinates_offset;
+    auto new_box_height = box_y1 - box_y0 + coordinates_offset;
+
+    // anchor_score: [1, 2 * num_anchors, feat_H, feat_W] -> [feat_H * feat_W * num_anchors, 2]
+    edsl::Tensor anchor_score = edsl::gather(class_probs, idx_zero + idx_batch).axis(0);
+    anchor_score = op::squeeze(anchor_score, {0});
+    anchor_score = edsl::reshape(anchor_score, {2, num_anchors, feat_H, feat_W});
+    anchor_score = op::transpose(anchor_score, edsl::make_tuple<int64_t>({2, 3, 1, 0}));
+    anchor_score = edsl::reshape(anchor_score, {num_proposals, 2});
+    // Currently only takes second backend scores referring to openvino implementation
+    anchor_score = edsl::gather(anchor_score, idx_one).axis(1);
+    auto valid_box_size = (new_box_width >= min_box_W) * (new_box_height >= min_box_H);
+    auto zero = edsl::cast(edsl::Tensor(0), anchor_score.dtype());
+    auto proposal_score = edsl::select(valid_box_size, anchor_score, zero);
+
+    auto batch_enum_proposals = op::concatenate({box_x0, box_y0, box_x1, box_y1, proposal_score}, 1);
+    batch_enum_proposals = op::unsqueeze(batch_enum_proposals, {0});
+    batch_proposals_vec.push_back(batch_enum_proposals);
+  }
+
+  // proposals: [num_batches, num_proposals, 5]
+  auto proposals = op::concatenate(batch_proposals_vec, 0);
+  return proposals;
+}
+
+edsl::Tensor partial_sort(edsl::Tensor proposals, int64_t pre_nms_topn, int64_t num_batches) {
+  edsl::Tensor idx_zero = edsl::index({edsl::TensorDim(1)}, 0);
+  edsl::Tensor idx_four = idx_zero + 4;
+
+  std::vector<edsl::Tensor> sorted_proposals_vec;
+  for (int64_t idx_batch = 0; idx_batch < num_batches; idx_batch++) {
```

Comment:
I wonder if you can use a 1D tensor for idx_batch to avoid the for loop.

Reply:
Done